lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/SharedState.hs

```diff
@@ -0,0 +1,357 @@
+-- CIP-1854 Multi-signature Wallets.
+--
+-- It indicates that the subtree of this node is used according to this
+-- specification.
+--
+-- Hardened derivation is used at this level.
+purposeCIP1854 :: Index 'Hardened 'PurposeK
+purposeCIP1854 = toEnum 0x8000073E
+
+-- | Create a new SharedState from public account key.
+mkSharedStateFromAccountXPub
+    :: SupportsSharedState n k
+    => k 'AccountK XPub
+    -> Index 'Hardened 'AccountK
+    -> AddressPoolGap
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> SharedState n k
+mkSharedStateFromAccountXPub accXPub accIx gap pTemplate dTemplateM =
+    mkSharedState accIx $
+    SharedStatePending accXPub pTemplate dTemplateM gap
+
+mkSharedState
+    :: SupportsSharedState n k
+    => Index 'Hardened 'AccountK
+    -> SharedStatePending k
+    -> SharedState n k
+mkSharedState accIx state = updateSharedState id $ SharedState
+    { derivationPrefix
+    , fields = PendingFields state
+    }
+  where
+    derivationPrefix = DerivationPrefix (purposeCIP1854, coinTypeAda, accIx)
+
+type SupportsSharedState (n :: NetworkDiscriminant) k =
+    ( MkKeyFingerprint k Address
+    , SoftDerivation k
+    , Typeable n
+    , MkKeyFingerprint k (Proxy n, k 'AddressK XPub)
+    )
+
+-- | Create a new SharedState from root private key and password.
+mkSharedStateFromRootXPrv
+    :: (SupportsSharedState n k, WalletKey k)
+    => (k 'RootK XPrv, Passphrase "encryption")
+    -> Index 'Hardened 'AccountK
+    -> AddressPoolGap
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> SharedState n k
+mkSharedStateFromRootXPrv (rootXPrv, pwd) accIx gap pTemplate dTemplateM =
+    mkSharedState accIx $
+    SharedStatePending accXPub pTemplate dTemplateM gap
+  where
+    accXPub = publicKey $ deriveAccountPrivateKey pwd rootXPrv accIx
+
+-- | Turn a "pending" into an "active" state or identity if already "active"
+updateSharedState
+    :: forall n k. SupportsSharedState n k
+    => (SharedStatePending k -> SharedStatePending k)
+    -> SharedState n k
+    -> SharedState n k
+updateSharedState f st = case fields st of
+    ReadyFields _ -> st
+    PendingFields pending -> case sharedStateFromPending @n (f pending) of
+        Just ready -> st { fields = ReadyFields ready }
+        Nothing -> st { fields = PendingFields (f pending) }
+
+sharedStateFromPending
+    :: forall n k. SupportsSharedState n k
+    => SharedStatePending k
+    -> Maybe (AddressPool 'MultisigScript k)
+sharedStateFromPending (SharedStatePending accXPub pT dT g)
+    | templatesComplete pT dT = Just $
+        mkAddressPool @n (ParentContextMultisigScript accXPub pT dT) g []
+    | otherwise = Nothing
+
+templatesComplete
+    :: ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> Bool
+templatesComplete pTemplate dTemplate =
+    isValid pTemplate && maybe True isValid dTemplate
+  where
+    isValid = isRight . validateScriptTemplate RequiredValidation
```

Comment:
:+1:

Reply:
On second thought, this is not sufficient to validate a template. We must also check that the top-level account key is present in the set of cosigners. Otherwise, the template is invalid. And by extension, we should have the following invariant always true (I'd be convinced by a test :) ):

- The payment (resp. delegation if defined) template must always contain at least one co-signer key binding that is equal to the top-level account key defined for shared wallet (that is, our key).

Also, it should not be possible to change the top-level account key associated with a shared wallet